contracts/src/v0.7/dev/KeeperRegistryDev.sol

```diff
@@ -674,50 +655,174 @@ contract KeeperRegistryDev is
       uint32 checkGasLimit,
       uint24 stalenessSeconds,
       uint16 gasCeilingMultiplier,
+      uint96 minUpkeepSpend,
+      uint32 maxPerformGas,
       uint256 fallbackGasPrice,
       uint256 fallbackLinkPrice,
-      uint32 maxPerformGas,
-      uint96 minUpkeepSpend
+      uint32 nonce
     )
   {
-    Config memory config = s_config;
+    State memory state = s_state;
     return (
-      config.paymentPremiumPPB,
-      config.blockCountPerTurn,
-      config.checkGasLimit,
-      config.stalenessSeconds,
-      config.gasCeilingMultiplier,
+      state.config.paymentPremiumPPB,
+      state.config.blockCountPerTurn,
+      state.config.checkGasLimit,
+      state.config.stalenessSeconds,
+      state.config.gasCeilingMultiplier,
+      state.config.minUpkeepSpend,
+      state.config.maxPerformGas,
       s_fallbackGasPrice,
       s_fallbackLinkPrice,
-      config.maxPerformGas,
-      config.minUpkeepSpend
+      state.nonce
     );
   }
 
-  /**
-   * @notice getFlatFee gets the flat rate fee charged to customers when performing upkeep,
-   * in units of of micro LINK
-   */
-  function getFlatFee() external view returns (uint32) {
-    return s_config.flatFeeMicroLink;
-  }
-
   /**
    * @notice calculates the minimum balance required for an upkeep to remain eligible
+   * @param id the upkeep id to calculate minimum balance for
    */
   function getMinBalanceForUpkeep(uint256 id) external view returns (uint96 minBalance) {
     return getMaxPaymentForGas(s_upkeep[id].executeGas);
   }
 
   /**
    * @notice calculates the maximum payment for a given gas limit
+   * @param gasLimit the gas to calculate payment for
    */
   function getMaxPaymentForGas(uint256 gasLimit) public view returns (uint96 maxPayment) {
     (uint256 gasWei, uint256 linkEth) = getFeedData();
     uint256 adjustedGasWei = adjustGasPrice(gasWei, false);
     return calculatePaymentAmount(gasLimit, adjustedGasWei, linkEth);
   }
 
+  /**
+   * @notice retrieves the permited incoming and outgoing registries from storage
+   */
+  function getPermittedRegistries() external view returns (address[] memory, address[] memory) {
+    uint256 idx;
+    address[] memory incoming = new address[](s_permittedIncomingRegistryAddresses.length());
+    address[] memory outgoing = new address[](s_permittedOutgoingRegistryAddresses.length());
+    for (; idx < incoming.length; idx++) {
+      incoming[idx] = s_permittedIncomingRegistryAddresses.at(idx);
+    }
+    for (idx = 0; idx < outgoing.length; idx++) {
+      outgoing[idx] = s_permittedOutgoingRegistryAddresses.at(idx);
+    }
+    return (incoming, outgoing);
+  }
+
+  /**
+   * @notice sets the approved registry addresses
+   * @dev this function wastes so much gas 🤦
```

Comment:
Another idea on saving space: Could we consolidate the permitted registry list into just a map? If could be a map that points to a uint, 0 is no permissions, 1 import, 2 export, 3 import+export.

I don't think we really need a readable list. We don't expect there to be a ton of these, and if you want to see the whole list without consulting a frontend somewhere you can query logs(don't really see the need to read this on-chain). I think we could get away with a single map, and a single setter for adding to the list.

Reply:
I think that's a great call and gas saver. Doing it.